### PR TITLE
Update sigs.k8s.io/structured-merge-diff to v4.4.0

### LINKED ISSUE #113482: Server-side apply fails with duplicate container ports and env-vars allowed in created/updated pods by validation
https://github.com/kubernetes/kubernetes/pull/113245#discussion_r1009911070

Validation allows:

```
        ports:
        - containerPort: 93
          hostPort: 9376
        - containerPort: 93
          hostPort: 9377
```

but server-side apply does not:

```
Error from server: failed to create manager for existing fields: failed to convert new object (default/hostnames; apps/v1, Kind=Deployment) to smd typed: .spec.template.spec.containers[name="serve-hostname-zp2lc"].ports: duplicate entries for key [containerPort=93,protocol="TCP"]
```

### PR DESCRIPTION
**Pull request by apelisse**
#### What type of PR is this?

/kind bug
/kind feature

It's a feature that solves a bug ...

#### Which issue(s) this PR fixes:
Fixes #113482

#### Does this PR introduce a user-facing change?
We probably need something here, not sure how to phrase it yet.
```release-note
Fixes bugs in handling of server-side apply, create, and update API requests for objects containing duplicate items in keyed lists.
* A `create` or `update` API request with duplicate items in a keyed list no longer wipes out managedFields. Examples include env var entries with the same name, or port entries with the same containerPort in a pod spec.
* A server-side apply request that makes unrelated changes to an object which has duplicate items in a keyed list no longer fails, and leaves the existing duplicate items as-is.
* A server-side apply request that changes an object which has duplicate items in a keyed list, and modifies the duplicated item removes the duplicates and replaces them with the single item contained in the server-side apply request.
```

### CODE REVIEW

**Comment by dims** [comment]
thanks @apelisse ! we had to jump through hoops in https://github.com/kubernetes/kubernetes/issues/113482#issuecomment-1772389117 glad this is sorted out.

/approve
/lgtm

**Comment by apelisse** [comment]
I might want to add specific tests but don't think I'll have time today.

**Comment by hailkomputer** [comment]
Hello, bug triage team lead here. I want to check the status and remind that the the code freeze is starting [01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023](https://everytimezone.com/s/24b27b03), as we want to ensure that each PR has a chance to be merged.
Are we still targeting the 1.29 release?

**Comment by liggitt** [comment]
shouldn't we have a test that exercises SSA with a duplicate port / duplicate envvar and a multi-fieldManager flow? something like:

1. create daemonset with duplicate port / duplicate envvar, capture managedFields
2. SSA unrelated fields as manager=A, make sure it succeeds, capture managedFields
3. update to modify the duplicate ports, make sure it succeeds and doesn't drop managedFields
4. SSA an unrelated port and envvar as manager=B, make sure it succeeds and doesn't drop duplicates, capture managedFields
5. SSA a single replacement for the duplicate port as manager=C, make sure it succeeds and drops duplicate, takes over managedField ownership for the port from the creater/updater

**Comment by liggitt** [comment]
(dependency update lgtm, would like to see a test walking through the scenario this is fixing and spitting out the resulting managedFields structures so we can verify an older server won't choke on those)

**Comment by apelisse** [comment]
OK, I needed to add the test and enable the AllowDuplicates option (thanks to the test for catching that).
From what I can tell, that behaves right as expected. Giving more thought to this comment:
> spitting out the resulting managedFields structures so we can verify an older server won't choke on those

I realized there is one case that smells:
What if we remove the duplicate on an old cluster. The cluster has to be new to create the duplicate. If we update the duplicate on an old server, it'll wipe the managed fields. But if we REMOVE the duplicates on an OLD server, then the old server will see no duplicate, and the weirdly formed managed field. I'll write a test in SMD to try and reproduce that.

**Comment by apelisse** [comment]
I'm fixing the smd tag though since v4.4.0 wasn't exactly where I wanted to do, updating to 4.4.1.

File: staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/structuredmerge.go [hunk 1]
```diff
@@ -71,6 +74,7 @@ func NewCRDStructuredMergeManager(typeConverter TypeConverter, objectConverter r
 			Converter:     newCRDVersionConverter(typeConverter, objectConverter, hub),
 			IgnoredFields: resetFields,
 		},
+		allowDuplicatesInSets: false,
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/structuredmerge.go:77]: CRs can have duplicates in sets if they were persisted prior to the schema being updated to mark a list a set, should this be true? should we have a test?

Comment (apelisse) [reply staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/structuredmerge.go:77]: Yeah that's a good point, we should probably allow it. The test won't add much though? it's basically the exact same path as built-ins.

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/structuredmerge.go:77]: ok, I didn't know if there were different paths taken for typed/structured vs untyped/openapi schema only

Comment (apelisse) [reply staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/structuredmerge.go]: I've updated it, it simplifies the code, I like it :-)

File: test/integration/apiserver/apply/apply_test.go [hunk 2]
```diff
@@ -3670,3 +3670,284 @@ func TestApplyFormerlyAtomicFields(t *testing.T) {
...
+		t.Fatalf("Expected %v Ports, got %v", want, got)
+	}
+	if want, got := 1, len(daemon.ManagedFields); want != got {
+		t.Fatalf("Expected %v ManagedFields, got %v", want, got)
+	}
+	m, err := json.Marshal(daemon.ManagedFields)
+	if err != nil {
+		t.Fatalf("Failed to serialize managed fields")
+	}
+	t.Logf("Managed fields after creation: %v", string(m))
```
[the remark this answers is not in the document]
Comment (apelisse) [reply test/integration/apiserver/apply/apply_test.go:3756]: Thanks, that looks good. Let me pull your commit.

[the remark this answers is not in the document]
Comment (apelisse) [reply test/integration/apiserver/apply/apply_test.go:3756]: > For [this bit](https://github.com/liggitt/kubernetes/commit/ab38418e6ba107124d8ed3e81688a4049cc11248#diff-b629323ef67f91837456deae39526a0e81aa0a5397465ff546e6845211afb13dR3795-R3800), I wasn't sure if the entry for the duplicate items should contain ".": {} or not... I can't remember what that means.

From what I can tell, it's fine. It's here for lists that can have an arbitrary length, I can't exactly remember how that works exactly :-(

[the remark this answers is not in the document]
Comment (apelisse) [reply test/integration/apiserver/apply/apply_test.go]: pushed.

**Comment by liggitt** [comment]
/lgtm
/approve

I ran a modified version of the integration test against master before this PR merges to ensure an old server doing an update of an object containing managedFields like this still works (it does, and still drops the managed fields), and an old server doing an apply of an object containing duplicates and managedFields like this does not fail worse (it doesn't, the apply still fails but only the immediate request fails).

**Comment by apelisse** [comment]
Thanks for the dilligence Jordan!

**Comment by thockin** [comment]
1) We DEFINTELY need a good release note.

2) Can someone summarize the change in terms of user-facing semantics?  We know several failure modes, are they totally eliminated?  Mitigated?  Compartmentalized?

**Comment by liggitt** [comment]
The failure modes I know of are eliminated. Updated the release note.

* A `create` or `update` API request with duplicate items in a keyed list no longer wipes out managedFields. Examples include env var entries with the same name, or port entries with the same containerPort in a pod spec.
* A server-side apply request that makes unrelated changes to an object which has duplicate items in a keyed list no longer fails, and leaves the existing duplicate items as-is.
* A server-side apply request that changes an object which has duplicate items in a keyed list, and modifies the duplicated item removes the duplicates and replaces them with the single item contained in the server-side apply request.

**Comment by apelisse** [comment]
Do we need to explain the actual impact to managed fields?

**Comment by liggitt** [comment]
> Do we need to explain the actual impact to managed fields?

I don't think so, that's pretty inside baseball to kube-apiserver

**Comment by thockin** [comment]
```
$ alias k=`pwd`/_output/bin/kubectl

$ cat /tmp/d.yaml 
apiVersion: apps/v1
kind: Deployment
metadata:
  name: ssa-test
spec:
  replicas: 1
  selector:
    matchLabels:
      app: ssa-test
  template:
    metadata:
      labels:
        app: ssa-test
    spec:
      containers:
      - name: c
        image: busybox
        ports:
        - containerPort: 93
          hostPort: 9376
        - containerPort: 93
          hostPort: 7693
        env:
        - name: FOO
          value: "foo"
        - name: FOO
          value: "foo-$(FOO)"

$ # With an older apiserver...

$ k delete -f /tmp/d.yaml; k create -f /tmp/d.yaml 
Error from server (NotFound): error when deleting "/tmp/d.yaml": deployments.apps "ssa-test" not found
Warning: spec.template.spec.containers[0].env[1].name: duplicate name "FOO"
deployment.apps/ssa-test created

$ k replace -f /tmp/d.yaml 
Warning: spec.template.spec.containers[0].env[1].name: duplicate name "FOO"
deployment.apps/ssa-test replaced

$ k apply -f /tmp/d.yaml 
Warning: resource deployments/ssa-test is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.
Warning: spec.template.spec.containers[0].env[1].name: duplicate name "FOO"
deployment.apps/ssa-test configured

$ k apply --server-side -f /tmp/d.yaml 
Error from server: failed to create manager for existing fields: failed to convert new object (default/ssa-test; apps/v1, Kind=Deployment) to smd typed: errors:
  .spec.template.spec.containers[name="c"].env: duplicate entries for key [name="FOO"]
  .spec.template.spec.containers[name="c"].ports: duplicate entries for key [containerPort=93,protocol="TCP"]

$ # With a brand-new apiserver (built minute ago from HEAD)...

$ k delete -f /tmp/d.yaml; k create -f /tmp/d.yaml 
deployment.apps "ssa-test" deleted
deployment.apps/ssa-test created

$ k replace -f /tmp/d.yaml 
deployment.apps/ssa-test replaced

$ k apply -f /tmp/d.yaml 
Warning: resource deployments/ssa-test is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.
deployment.apps/ssa-test configured

$ k apply --server-side -f /tmp/d.yaml 
Error from server: failed to create manager for existing fields: failed to convert new object (default/ssa-test; apps/v1, Kind=Deployment) to smd typed: errors:
  .spec.template.spec.containers[name="c"].env: duplicate entries for key [name="FOO"]
  .spec.template.spec.containers[name="c"].ports: duplicate entries for key [containerPort=93,protocol="TCP"]
```

On the new apiserver the "create" operation logged:

```
"[SHOULD NOT HAPPEN] failed to update managedFields" err=<
	failed to convert new object (default/ssa-test; apps/v1, Kind=Deployment) to smd typed: errors:
	  .spec.template.spec.containers[name="c"].env: duplicate entries for key [name="FOO"]
	  .spec.template.spec.containers[name="c"].ports: duplicate entries for key [containerPort=93,protocol="TCP"]
 > versionKind="/, Kind=" namespace="default" name="ssa-test"
```

On the SSA operation (note there's no change in the file) it logged:

```
apiserver received an error that is not an metav1.Status: &errors.errorString{s:"failed to create manager for existing fields: failed to convert new object (default/ssa-test; apps/v1, Kind=Deployment) to smd typed: errors:\n  .spec.template.spec.containers[name=\"c\"].env: duplicate entries for key [name=\"FOO\"]\n  .spec.template.spec.containers[name=\"c\"].ports: duplicate entries for key [containerPort=93,protocol=\"TCP\"]"}: failed to create manager for existing fields: failed to convert new object (default/ssa-test; apps/v1, Kind=Deployment) to smd typed: errors:
  .spec.template.spec.containers[name="c"].env: duplicate entries for key [name="FOO"]
  .spec.template.spec.containers[name="c"].ports: duplicate entries for key [containerPort=93,protocol="TCP"]
```

**Comment by apelisse** [comment]
You still can't server-side apply duplicates.

**Comment by thockin** [comment]
Apparently I screwed up my rebase and didn't have this change.  Let me try again

**Comment by thockin** [comment]
```
$ k delete -f /tmp/d.yaml; k create -f /tmp/d.yaml 
deployment.apps "ssa-test" deleted
deployment.apps/ssa-test created

$ k replace -f /tmp/d.yaml 
deployment.apps/ssa-test replaced

$ k apply -f /tmp/d.yaml 
Warning: resource deployments/ssa-test is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.
deployment.apps/ssa-test configured

$ k apply --server-side -f /tmp/d.yaml 
Error from server: failed to create typed patch object (default/ssa-test; apps/v1, Kind=Deployment): errors:
  .spec.template.spec.containers[name="c"].env: duplicate entries for key [name="FOO"]
  .spec.template.spec.containers[name="c"].ports: duplicate entries for key [containerPort=93,protocol="TCP"]

$ cat /tmp/d2.yaml 
apiVersion: apps/v1
kind: Deployment
metadata:
  name: ssa-test
  labels:
    foo: bar

$ k apply --server-side -f /tmp/d2.yaml 
deployment.apps/ssa-test serverside-applied
```

As my teenager would say... "my bad"

**Comment by thockin** [comment]
I do still get logs:

```
E1101 17:51:35.316459 3700183 status.go:71] apiserver received an error that is not an metav1.Status: &errors.errorString{s:"failed to create typed patch object (default/ssa-test; apps/v1, Kind=Deployment): errors:\n  .spec.template.spec.containers[name=\"c\"].env: duplicate entries for key [name=\"FOO\"]\n  .spec.template.spec.containers[name=\"c\"].ports: duplicate entries for key [containerPort=93,protocol=\"TCP\"]"}: failed to create typed patch object (default/ssa-test; apps/v1, Kind=Deployment): errors:
  .spec.template.spec.containers[name="c"].env: duplicate entries for key [name="FOO"]
  .spec.template.spec.containers[name="c"].ports: duplicate entries for key [containerPort=93,protocol="TCP"]
```

**Comment by apelisse** [comment]
Something/someone is sending a patch with duplicates. That's literally trying to parse the server-side applied object.

**Comment by liggitt** [comment]
yeah, you can't SSA duplicate items, but you can SSA to objects which have duplicate items added via create or update calls
